tracer/src/Datadog.Trace/Propagators/W3CBaggagePropagator.cs

```diff
@@ -0,0 +1,270 @@
+// </copyright>
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+using Datadog.Trace.Telemetry;
+using Datadog.Trace.Telemetry.Metrics;
+using Datadog.Trace.Util;
+
+#nullable enable
+
+namespace Datadog.Trace.Propagators;
+
+// https://www.w3.org/TR/baggage/
+internal class W3CBaggagePropagator : IContextInjector, IContextExtractor
+{
+    public const string BaggageHeaderName = "baggage";
+    public const int DefaultMaximumBaggageItems = 64;
+    public const int DefaultMaximumBaggageBytes = 8192;
+
+    // the standard W3C separator between key/value pairs
+    // "key1=value1,key2=value2"
+    //             ^
+    private const char PairSeparator = ',';
+
+    // the standard W3C separator between the key and value in each pair
+    // "key1=value1,key2=value2"
+    //      ^           ^
+    private const char KeyAndValueSeparator = '=';
+
+    public static readonly W3CBaggagePropagator Instance = new();
+
+    private static readonly HashSet<char> KeyCharsToEncode;
+
+    private static readonly HashSet<char> ValueCharsToEncode;
+
+    static W3CBaggagePropagator()
+    {
+        // key may not contain whitespace or any of the following characters: " , ; \ ( ) / : < = > ? @ [ ] { }
+        KeyCharsToEncode = ['"', ',', ';', '\\', '(', ')', '/', ':', '<', '=', '>', '?', '@', '[', ']', '{', '}'];
+
+        // value may contain characters from the Basic Latin Unicode Block, except for the following:
+        //   U+0020 space ( )
+        //   U+0022 double quotation mark (")
+        //   U+002C comma (,)
+        //   U+003B semicolon (;)
+        //   U+005C backslash (\)
+        ValueCharsToEncode = [' ', '"', ',', ';', '\\'];
+    }
+
+    private W3CBaggagePropagator()
+    {
+    }
+
+    public PropagatorType PropagatorType => PropagatorType.Baggage;
+
+    public void Inject<TCarrier, TCarrierSetter>(
+        PropagationContext context,
+        TCarrier carrier,
+        TCarrierSetter carrierSetter)
+        where TCarrierSetter : struct, ICarrierSetter<TCarrier>
+    {
+        var baggage = context.Baggage;
+
+        if (baggage is null or { Count: 0 })
+        {
+            // nothing to inject
+            return;
+        }
+
+        GetSettings(context, out var baggageMaximumItems, out var baggageMaximumBytes);
+        var headerValue = CreateHeader(baggage, baggageMaximumItems, baggageMaximumBytes);
+
+        if (!string.IsNullOrWhiteSpace(headerValue))
+        {
+            carrierSetter.Set(carrier, BaggageHeaderName, headerValue);
+            TelemetryFactory.Metrics.RecordCountContextHeaderStyleInjected(MetricTags.ContextHeaderStyle.Baggage);
+        }
+    }
+
+    private static void GetSettings(PropagationContext context, out int baggageMaximumItems, out int baggageMaximumBytes)
+    {
+        if (context.SpanContext?.TraceContext?.Tracer.Settings is { } settings)
```

Comment:
I would like to avoid having to look up the tracer on each of these invocations. Can we pass the settings into the SpanContextPropagatorFactory so we can initialize the propagator instance with these settings?

Reply:
There's another issue here, too: if we're propagating baggage without an active trace, then `context.SpanContext` can be `null` and we currently have no way of accessing the settings from here, which is why I'm falling back to the defaults for now.

Which reminds me, we should make the http client instrumentation inject context even if the integration itself is disabled and we don't generate a span.

But back to your comment, currently we create all the propagators only once as singletons and `SpanContextPropagatorFactory` simply returns `W3CBaggage.Instance`. If we pass settings to the propagator ctor in `SpanContextPropagatorFactory` and the settings ever changed, the (singleton) baggage propagator would not be updated.

Another option could be to access `Tracer.Instance.Setting` here, I generally try to avoid accessing it that way.

Reply:
Hmm do we limits that differ from the [defaults](https://w3c.github.io/baggage/#limits)? If we do, then maybe we should avoid returning `W3CBaggage.Instance` in the `SpanContextPropagatorFactory` class. Maybe we can return that only if the settings are the default ones?

Reply:
Our defaults are the same as the W3C recommended limits, but we allow users to override if needed. We need those values when generating the header for injection. One way is to pass then into the `W3CBaggage` ctor, which means that propagator now has some state, and we would need to create a new propagator instance at least every time the settings change. Another way would be to keep things as stateless as possible, and pass these value into the `Inject` methods instead, but none of our propagators do that today since they all implemented `IContextInjector`.

Given the `Inject` arguments (`context`, `carrier`, and `carrierSetter`), this was the only way to access them through `context`. The other option is to go to `Tracer.Instance` directly, which is probably fine, and actually worked even when `context.SpanContext` is null. We can even combine them those two as:
```csharp
var tracer = context.SpanContext?.TraceContext?.Tracer ?? Tracer.Instance;
var settings = tracer.Settings;
```

Reply:
Gotcha. I'd be open to adding state to the propagator and newing one up each time we update settings, but if you'd rather keep the existing instances then I'd prefer only going to `Tracer.Instance` and not trying to access the tracer settings through the SpanContext

Reply:
> I'd be open to adding state to the propagator and newing one up each time we update settings

I looked into this, but it involves a bit more refactoring that we can do later if we want. For now I changed this to get the settings from `Tracer.Instance` instead.